File: resb/src/resb/dp.py
"""Differential privacy utilities."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict

import numpy as np
import pandas as pd


@dataclass(frozen=True)
class DPReport:
    """Summary of the DP configuration used to perturb synthetic samples."""

    epsilon: float
    delta: float
    sigma: Dict[str, float]


def gaussian_sigma(epsilon: float, delta: float, sensitivity: float) -> float:
    """Calibrate the Gaussian mechanism standard deviation."""

    if sensitivity <= 0:
        raise ValueError("sensitivity must be positive")
    if epsilon <= 0:
        raise ValueError("epsilon must be positive")
    if not (0 < delta < 1):
        raise ValueError("delta must be within (0, 1)")
    return float(np.sqrt(2 * np.log(1.25 / delta)) * sensitivity / epsilon)


def infer_sensitivity(minority: pd.DataFrame) -> Dict[str, float]:
    """Estimate feature sensitivities from the minority slice."""

    sensitivities: Dict[str, float] = {}
    for column in minority.select_dtypes(include=["number"]).columns:
        col = minority[column]
        sensitivities[column] = float(col.max() - col.min() or 1.0)
    return sensitivities


def apply_gaussian_noise(
    df: pd.DataFrame,
    sigma_map: Dict[str, float],
    rng: np.random.Generator,
) -> pd.DataFrame:
    """Apply Gaussian noise to numeric columns as per ``sigma_map``."""

    noisy = df.copy()
    for column, sigma in sigma_map.items():
        if sigma <= 0:
            continue
        if column not in noisy:
            continue
        if not np.issubdtype(noisy[column].dtype, np.number):
            continue
        noise = rng.normal(loc=0.0, scale=sigma, size=len(noisy))
        noisy[column] = noisy[column] + noise
    return noisy


def validate_dp_parameters(epsilon: float, delta: float) -> None:
    if epsilon <= 0:
        raise ValueError("epsilon must be positive")
    if not (0 < delta < 1):
        raise ValueError("delta must be within (0, 1)")

File: server/src/dp.ts
export type DpConfig = {
  epsilon: number;
  delta?: number;
  sensitivity: number;
  mechanism: 'laplace' | 'gaussian';
};

function randomNormal(mu = 0, sigma = 1) {
  // Box-Muller
  let u = 0,
    v = 0;
  while (u === 0) u = Math.random();
  while (v === 0) v = Math.random();
  return (
    mu + sigma * Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v)
  );
}

export function dpNoise(cfg: DpConfig) {
  if (cfg.mechanism === 'laplace') {
    const b = cfg.sensitivity / cfg.epsilon;
    return (Math.random() - 0.5) * 2 * b;
  }
  const sigma =
    Math.sqrt(2 * Math.log(1.25 / (cfg.delta || 1e-6))) *
    (cfg.sensitivity / cfg.epsilon);
  return randomNormal(0, sigma);
}

export function applyDp(count: number, cfg: DpConfig) {
  return count + dpNoise(cfg);
}
